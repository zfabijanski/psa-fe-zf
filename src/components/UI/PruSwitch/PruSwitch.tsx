import React, { ChangeEvent } from "react";
import styled from "styled-components/macro";

export interface PruSwitchProps
  extends React.InputHTMLAttributes<HTMLInputElement> {
  onValueChange?: (checked: boolean) => void;
}

export const PruSwitch = React.forwardRef<HTMLInputElement, PruSwitchProps>(
  ({ onValueChange, ...rest }, ref) => {
    const handleValueChange = (e: ChangeEvent<HTMLInputElement>) => {
      if (rest.onChange) {
        rest.onChange(e);
      }
      if (onValueChange) {
        onValueChange(e.target.checked);
      }
    };
    return (
      <Switch className={rest.className} role="checkbox">
        <Input {...rest} type="checkbox" onChange={handleValueChange} />
        <Slider />
      </Switch>
    );
  }
);

const Switch = styled.label`
  position: relative;
  width: 44px;
  min-width: 44px;
  height: 24px;
  display: block;
`;

const Slider = styled.div`
  font-family: "pru-internal_iconfont" !important;
  text-align: center;
  font-weight: bold;
  cursor: pointer;
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  top: 0;
  transition: 0.2s;
  border-radius: 11.5px;
  border: 1.5px solid ${({ theme }) => theme.newColors.gray100};
  background-color: ${({ theme }) => theme.newColors.white100};

  &:before {
    content: "\ue947";
    color: ${({ theme }) => theme.newColors.white100};
    background-color: ${({ theme }) => theme.newColors.gray100};
    line-height: 24px;
    height: 24px;
    width: 24px;
    position: absolute;
    top: -1.5px;
    left: -1.5px;
    transition: 0.2s;
    border-radius: 50%;
  }
`;

const Input = styled.input`
  display: none;

  &:checked + ${Slider} {
    border: none;
    background-color: ${({ theme }) => theme.newColors.primary100};

    &:before {
      content: "\ue90f";
      color: ${({ theme }) => theme.newColors.gray100};
      background-color: ${({ theme }) => theme.newColors.white100};
      height: 20px;
      width: 20px;
      line-height: 21px;
      position: absolute;
      top: 2px;
      left: 22px;
      transition: 0.2s;
      border-radius: 50%;
    }
  }
`;

export default PruSwitch;
