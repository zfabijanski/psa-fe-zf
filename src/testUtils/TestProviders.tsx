import { render, RenderOptions } from "@testing-library/react";
import { FC, PropsWithChildren, ReactElement } from "react";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";
import { Normalize } from "styled-normalize";
import GlobalStyles from "theme/globalStyles";
import { theme } from "theme/theme";
import rootStore, { configureStore, rootReducer as reducer } from "AppStore";
import { BrowserRouter } from "react-router-dom";
import { Modal } from "components/UI/Modal";
import { IntlProvider } from "react-intl";
import { plMessages } from "translations/pl";
import I18nProvider from "I18nProvider";

interface TestProvidersProps {
  store?: typeof rootStore;
  renderModal?: boolean;
  realIntlProvider?: boolean;
}

export const addTestProviders =
  ({
    store = configureStore({ reducer }),
    renderModal = true,
    // @TODO: update tests to use realIntlProvider everywhere
    realIntlProvider = false,
  }: TestProvidersProps): FC<PropsWithChildren<{}>> =>
  ({ children }) => {
    const TranslationsProvider = realIntlProvider ? IntlProvider : I18nProvider;

    return (
      <Provider store={store}>
        <BrowserRouter>
          <ThemeProvider theme={theme}>
            <TranslationsProvider locale="pl" messages={plMessages}>
              <Normalize />
              <GlobalStyles />
              {children}
              {renderModal && <Modal />}
            </TranslationsProvider>
          </ThemeProvider>
        </BrowserRouter>
      </Provider>
    );
  };

const customRender = (
  ui: ReactElement,
  options: Omit<RenderOptions, "options"> = {},
  testProviderOptions: TestProvidersProps = {}
) => render(ui, { wrapper: addTestProviders(testProviderOptions), ...options });

export * from "@testing-library/react";

// override render method
export { customRender as render };
