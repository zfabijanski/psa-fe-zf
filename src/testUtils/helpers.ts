export const sleep = (ms: number = 100): Promise<unknown> =>
  new Promise((resolve) => setTimeout(resolve, ms));
