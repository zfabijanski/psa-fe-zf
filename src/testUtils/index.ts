export * from "./TestProviders";
export * from "./mockCreators";
export * from "./helpers";
