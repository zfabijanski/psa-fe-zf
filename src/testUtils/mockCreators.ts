import {
  ICalculatorCoversConfig,
  IProductCovers,
} from "features/Calculator/types";
import {
  DocumentState,
  IDirectory,
  IDirectoryMetadata,
  IDocumentMetadata,
} from "features/Library/types";
import { IMeeting } from "features/Meeting/History/types";
import { IIllustration, ProductGuid } from "features/Products/types";
import { IModalData, ModalTypes } from "models/Modal";
import {
  Frequency,
  IFundRiskProfileConfig,
  IProductConfig,
  IProductFund,
  IProductGuarantee,
  InvestmentProfile,
  Role,
} from "models/common";
import {
  CalculationStatus,
  Direction,
  ICalculationDetailsDto,
  ICover,
  IInsured,
  YesNo,
} from "models/calculator";

export const createCalculationDetailsDto = (
  overrides?: Partial<ICalculationDetailsDto>
): ICalculationDetailsDto => ({
  calculation_id: 1,
  working_date: "2020-01-01",
  premiums_total: 100,
  guaranteed_death_benefit: 100,
  guaranteed_maturity_benefit: 100,
  projected_maturity_value_egr_m: 100,
  product_categories: [],
  uuid: "uuid",
  version: "version",
  source_system: "source_system",
  product_guid: 1,
  insureds_list: [],
  covers_list: [],
  funds_list: [],
  ...overrides,
});

export const createInsured = (overrides?: Partial<IInsured>): IInsured => ({
  insured_idx: 1,
  insured_uuid: "insured_uuid",
  age_in_months: 1,
  date_of_birth: "2020-01-01",
  role: Role.MAIN,
  ...overrides,
});

export const createCover = (overrides?: Partial<ICover>): ICover => ({
  cover_id: 1,
  cover_idx: 1,
  insured_idx: 1,
  code: "code",
  frequency_code: Frequency.Annually,
  payments: 1,
  direction_of_calculation: Direction.SumAssuredToPremium,
  investment_profile: InvestmentProfile.Balanced,
  indexation: YesNo.Yes,
  sum_assured: 1,
  premium_per_freq: 1,
  duration: 1,
  idx_option: YesNo.Yes,
  rop_option: YesNo.Yes,
  add_sum_assured: 1,
  ending_age: 1,
  exp_mat_ben: 1,
  ...overrides,
});

export const createCalculatorCoversConfig = (
  overrides?: Partial<ICalculatorCoversConfig>
): ICalculatorCoversConfig => ({
  coverName: "coverName",
  isChild: false,
  isHospital: false,
  ...overrides,
});

export const createCalculatorCoversConfigMap = (
  overrides?: Partial<Record<string, ICalculatorCoversConfig>>
): Record<string, ICalculatorCoversConfig> => ({
  coverName1: createCalculatorCoversConfig({ coverName: "coverName1" }),
  coverName2: createCalculatorCoversConfig({ coverName: "coverName2" }),
  ...overrides,
});

export const createProductCovers = (
  overrides?: Partial<IProductCovers>
): IProductCovers => ({
  mainCover: "mainCover",
  additionalCovers: ["additionalCover1", "additionalCover2"],
  wopCovers: ["wopCover1", "wopCover2"],
  childCovers: ["childCover1", "childCover2"],
  ...overrides,
});

export const createProductCoversMap = (
  overrides?: Partial<Record<string, IProductCovers>>
): Record<string, IProductCovers> => ({
  mainCover1: createProductCovers({ mainCover: "mainCover1" }),
  mainCover2: createProductCovers({ mainCover: "mainCover2" }),
  ...overrides,
});

export const createProductConfig = (
  overrides?: Partial<IProductConfig>
): IProductConfig => ({
  product_guid: 1,
  product_name: "product_name",
  product_categories: [],
  product_freqs: [],
  product_covers: [],
  product_funds: [],
  ...overrides,
});

export const createProductConfigMap = (
  overrides?: Partial<Record<string, IProductConfig>>
): Record<string, IProductConfig> => ({
  1: createProductConfig({ product_name: "product_name1" }),
  2: createProductConfig({ product_name: "product_name2" }),
  ...overrides,
});

export const createFundRiskProfileConfig = (
  overrides?: Partial<IFundRiskProfileConfig>
): IFundRiskProfileConfig => ({
  code: "code",
  name_pl: "name_pl",
  name_en: "name_en",
  risk_level: 1,
  status_code: "status_code",
  notes: "notes",
  ...overrides,
});

export const createProductGuarantee = (
  overrides?: Partial<IProductGuarantee>
): IProductGuarantee => ({
  guarantee_code: "guarantee_code",
  guarantee_name: "guarantee_name",
  guarantee_min_value: "1",
  guarantee_duration_months: "4",
  guarantee_percentages: [{ guarantee_percent: 10 }],
  ...overrides,
});

export const createProductFund = (
  overrides?: Partial<IProductFund>
): IProductFund => ({
  cover_code: "cover_code",
  fund_code: "fund_code",
  fund_name: "fund_name",
  fund_risk_profile: "code",
  fund_order: 1,
  guarantees: [createProductGuarantee()],
  ...overrides,
});

export const createModalData = (
  overrides: Partial<IModalData> = {}
): IModalData => ({
  modalIllustration: "alert-fill",
  modalTitleTrKey: "modalTitleTrKey",
  modalContentTrKey: "modalContentTrKey",
  modalContent: "modalContent",
  modalButtons: {
    confirm: { textTrKey: "modal.ok" },
  },
  modalType: ModalTypes.default,
  maskClosable: true,
  ...overrides,
});

export const createIllustration = (
  overrides: Partial<IIllustration> = {}
): IIllustration => ({
  idd_id: 1,
  product_guid: ProductGuid.Savings,
  adequacy_id: 1,
  order_id: 1,
  expected_amount: { value: 3 },
  premium_total: { value: 3 },
  duration: { value: 3 },
  premium_main_cover: { value: 3 },
  frequency: { value: 3 },
  sum_of_premiums: { value: 3 },
  sum_assured: { value: 3 },
  investment_profile: { value: 3 },
  pru_calc_calculation_id: 1,
  status: CalculationStatus.Illustrated,
  json_needs_analysis: null,
  ...overrides,
});

export const createDirectory = (
  overrides?: Partial<IDirectory>
): IDirectory => ({
  directoryId: 1,
  directories: [],
  documentsIds: [1, 2, 3],
  ...overrides,
});

export const createDirectoryMetadata = (
  overrides?: Partial<IDirectoryMetadata>
): IDirectoryMetadata => ({
  directoryId: 1,
  name: "name",
  isOpen: false,
  ...overrides,
});

export const createDocumentMetadata = (
  overrides?: Partial<IDocumentMetadata>
): IDocumentMetadata => ({
  documentId: 1,
  guid: "guid",
  name: "name",
  state: DocumentState.NotAttached,
  isStateChanging: false,
  moduleName: "PRU_LIBRARY",
  ...overrides,
});

export const createMeeting = (overrides?: Partial<IMeeting>): IMeeting => ({
  meetingNo: "meetingNo",
  meetingId: 1,
  name: "name",
  age: 18,
  lastAccessDate: new Date("2023-01-01"),
  createDate: new Date("2023-01-01"),
  statusId: 1,
  statusName: "statusName",
  statusDescription: "statusDescription",
  state: "state",
  haveAdequacy: false,
  haveCalculation: false,
  haveIdd: false,
});
