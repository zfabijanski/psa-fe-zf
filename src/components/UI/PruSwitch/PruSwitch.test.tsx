import { PruSwitch, PruSwitchProps } from "./PruSwitch";
import { render, screen, fireEvent } from "testUtils";

const mockOnValueChange = jest.fn();

const defaultProps: PruSwitchProps = {
  onValueChange: mockOnValueChange,
};

const setup = (props: Partial<PruSwitchProps> = {}) =>
  render(<PruSwitch {...defaultProps} {...props} />);

describe("PruSwitch", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly", () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("should call onValueChange when switch is clicked", () => {
    setup();
    fireEvent.click(screen.getByRole("checkbox"));
    expect(mockOnValueChange).toHaveBeenCalledTimes(1);
  });

  it("should call custom change handler when switch is clicked", () => {
    const mockCustomChangeHandler = jest.fn();
    setup({ onChange: mockCustomChangeHandler });
    fireEvent.click(screen.getByRole("checkbox"));
    expect(mockCustomChangeHandler).toHaveBeenCalledTimes(1);
    expect(mockOnValueChange).toHaveBeenCalledTimes(1);
  });

  it("should allow to onValueChange to be undefined", () => {
    setup({ onValueChange: undefined });
    fireEvent.click(screen.getByRole("checkbox"));
    expect(mockOnValueChange).toHaveBeenCalledTimes(0);
  });
});
