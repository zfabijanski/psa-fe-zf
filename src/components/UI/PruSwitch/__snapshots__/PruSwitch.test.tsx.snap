// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PruSwitch should render correctly 1`] = `
.c0 {
  position: relative;
  width: 44px;
  min-width: 44px;
  height: 24px;
  display: block;
}

.c3 {
  font-family: "pru-internal_iconfont" !important;
  text-align: center;
  font-weight: bold;
  cursor: pointer;
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  top: 0;
  -webkit-transition: 0.2s;
  transition: 0.2s;
  border-radius: 11.5px;
  border: 1.5px solid #2F2F2F;
  background-color: #FFFFFF;
}

.c3:before {
  content: "";
  color: #FFFFFF;
  background-color: #2F2F2F;
  line-height: 24px;
  height: 24px;
  width: 24px;
  position: absolute;
  top: -1.5px;
  left: -1.5px;
  -webkit-transition: 0.2s;
  transition: 0.2s;
  border-radius: 50%;
}

.c1 {
  display: none;
}

.c1:checked + .c2 {
  border: none;
  background-color: #055A60;
}

.c1:checked + .c2:before {
  content: "";
  color: #2F2F2F;
  background-color: #FFFFFF;
  height: 20px;
  width: 20px;
  line-height: 21px;
  position: absolute;
  top: 2px;
  left: 22px;
  -webkit-transition: 0.2s;
  transition: 0.2s;
  border-radius: 50%;
}

<div>
  <label
    class="c0"
    role="checkbox"
  >
    <input
      class="c1"
      type="checkbox"
    />
    <div
      class="c2 c3"
    />
  </label>
  <div />
</div>
`;
